### DISCUSSION TITLE
FSharp.Html ?

### DISCUSSION THREAD

**Discussion by jkone27** [dotnet/fsharp#16124] [Ideas]
could be useful and cool to have an HTML templating library baked into F# lang?

reason: many view engines are available, mainly in 3 styles:

### 1. List higher order f.
- Giraffe/Fable/Falco/Saturn view engine (list higher-order functions) : https://github.com/giraffe-fsharp/Giraffe.ViewEngine
- Feliz view engine (list higher-order functions more succint): https://github.com/dbrattli/Feliz.ViewEngine

### 2. CEs/stateful/reactive
- CEs view engines: https://fsbolero.io/docs/HTML , Fable.Vide (also integrates Vide state reactive lib so is a special case)

### 3. HTML string templates with interpolation $
- fable lit / Fun.blazor ,templated HTML string $: https://fable.io/Fable.Lit/docs/templates.html 

## Unification of (1. higher order f.)
* pros: could all view engines and other projects build on top of that? then changes would be in a single place

* cons: complex, cost. area of compatibility and maybe need to have different modules, for different scopes?

**Comment by vzarytovskii** [comment]
Probably not in the standard library/compiler. It's too specific, would be great if community will come up with something, and we can host and amplify it as part of fsprojects

**Reply by AngelMunoz** [reply]
I agree with you that this kind of library may be too specific to be in the core lib/compiler

But I think there's a C# feature that could help in this area:

The Interpolated String Handlers:
https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/interpolated-string-handler#implement-the-handler-pattern

They could enable string based templating engines as well as other kinds of text DSLs

In the case of Fable and friends I'd think they could enable a feature like [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates) and example of that feature's usage can be found in [this list](https://github.com/kay-is/awesome-tagged-templates)

I intended to open a suggestion to add Interpolated String Handlers in F# but I always forget it.

**Reply by edgarfgp** [reply]
There seems to be an open suggestion for `InterpolatedStringHandler` https://github.com/fsharp/fslang-suggestions/issues/1108

**Reply by AngelMunoz** [reply]
Oh nice, I had not seen it, thanks @edgarfgp

**Reply by Xyncgas** [reply]
You can use blazor in F# technically, it's possible for efficient implementation to be existing somewhere, you can simply build components like you would coding in blazor to build a website that has html codes

**Reply by Xyncgas** [reply]
By the way you already can find html library in .NET, if we are not introducing a way to build html easily into the language then we are not talking about changing the language, so there are libraries we can use, and if the libraries aren't too functional or F# flavored, we certainly can start with porting those library to using Compute-Expression in F#

I bet it would be cool if you can control in your F# project how your website's theme would look like, by theming it using project wide globals
(For example, as a lord in my domain because I am a programmer, I give 9 different names to div, each can be themed, say div2 always zooms in 90% of its parent's size etc, but of course, you can overwrite div using css as normal)

I bet it would be cool if you can build web ui component without using blazor, and consume the components in other fsharp solutions, and you can put a lot of things and do a lot of things with the these compute expressions so you can build something that can serialized to utf-8 html string or as bytes or as custom binary format to control an application's UI locally or remotely so you can truly enable scenarios of running high performance cloud application in cheap laptop / cell phone because the UI is rendered in the cloud and instead of streaming video it streams from the UI scripting engine in the cloud powered by fsharp.html so it can achieve higher fps and since we are doing this might as well invent our own UI format instead of the 'high-overhead' html as the new standard of Ui-Descripting-Language and let fsharp.html interop with fsharp.udl
